src/org/mozilla/javascript/optimizer/Optimizer.java

```diff
@@ -303,6 +303,8 @@ private int rewriteForNumberVariables(Node n, int desired) {
                 }
 
             case Token.BITNOT:
+            case Token.POS:
+            case Token.NEG:
                 {
```

Comment:
I've never looked at this area of the code before, but should BITNOT, POS, and NEG be moved up to the same section as INC and DEC, since they are all unary operators?

I could be wrong, because I've been looking at this for a while and still can't figure out how it works, but intuitively, it seems like these should all be the same.

Reply:
I realized the difference may be in that INC and DEC mutate the variable, but I still don't understand how this works :)

Reply:
I used binary operators such as `Token.SUB` and unary operators such as `Token.INC` as a reference here.

I'm not sure if `NoType` or `AnyType` is the right choice here.
https://github.com/mozilla/rhino/blob/ae3563a99b88585d8d4ee72ca03466a33a58dd79/src/org/mozilla/javascript/optimizer/Optimizer.java#L316

Reply:
From what I can tell, in the case where `(child.getType() == Token.GETELEM || child.getType() == Token.GETPROP)` then type will always be NoType, so that extra check in INC and DEC seems unnecessary. Though, I wonder if it _should_ be possible for them to return something else.

For example in the INC case, I would think if there ever were a case where the child.getType() is GETELEM and type is NumberType, it would be for something like:
```js
(function() {
  var a = [1,2];
  a[0]++;
})()
```

I do feel like I'm in a bit over my head here, though, since I don't understand exactly how everything is supposed to work. I don't suppose anyone knows of any design docs or references that explain the optimizer?